File: hw1/python/task2.py
import sys
import json
from pyspark import SparkContext

def test(sc):
    rdd = sc.textFile('review.json').map(json.loads)
    print(rdd.take(3))

def t2_spark(review_file, business_file, n):
    sc = SparkContext("local[*]", 'task2')
    rdd_review = sc.textFile(review_file).map(json.loads).map(lambda x: [x['business_id'], x['stars']])
    rdd_review_1 = rdd_review.filter(lambda x: x[1] is not None).groupByKey().mapValues(list)
    rdd_business = sc.textFile(business_file).map(json.loads).map(lambda x: (x['business_id'], x['categories']))
    rdd_business_1 = rdd_business.filter(lambda x: x[1] is not None).filter(lambda x: x[1].strip() != "")
    rdd = rdd_business_1.join(rdd_review_1).map(lambda x:(x[1][0],x[1][1])).flatMap(lambda x: [[c.strip(), x[1]] for c in x[0].split(',')])
    rdd2 = rdd.reduceByKey(lambda x, y: x + y).mapValues(lambda x:sum(x)/len(x))
    rdd3 = rdd2.sortBy(lambda x: x[0]).sortBy(lambda x: x[1], ascending=False)

    return rdd3.map(list).take(n)

def t2_no_spark(review_file, business_file, n):
    review_dict = dict()
    with open(review_file, 'r') as f:
        line = f.readline()
        while line:
            item = json.loads(line)
            if item['stars'] is not None:
                review_dict.setdefault(item['business_id'], []).append(item['stars'])
            line = f.readline()
    business_dict = dict()
    with open(business_file, 'r') as f:
        line = f.readline()
        while line:
            item = json.loads(line)
            if item['categories'] is not None and item['categories'].strip() != '':
                business_dict[item['business_id']] = item['categories']
            line = f.readline()

    join_dict = dict()
    for k in business_dict.keys():
        if k in review_dict.keys():
            categories = business_dict[k].strip().split(',')
            for c in categories:
                for s in review_dict[k]:
                    join_dict.setdefault(c.strip(), []).append(s)
    avg = list(map(lambda x: [x[0], sum(x[1])/len(x[1])], join_dict.items()))
    tmp = sorted(avg, key=lambda x:x[0])
    result = sorted(tmp, key=lambda x:x[1], reverse=True)[:n]
    return result


if __name__ == '__main__':

    # review_file, business_file, output_file, if_spark, n = "review.json", "business.json", 'task2_spark_ans', 'spark', 10
    review_file, business_file, output_file, if_spark, n = sys.argv[1:]
    n = int(n)
    if if_spark.strip() == 'spark':
        result = t2_spark(review_file, business_file, n)
    elif if_spark.strip() == 'no_spark':
        result = t2_no_spark(review_file, business_file, n)

    with open(output_file, 'w') as f:
        json.dump({"result": result}, f)

File: hw1/scala/task2.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import scala.io.Source
import scala.util.parsing.json.JSON

object task2 {
  def t2_spark(review_file: String, business_file: String, output_file: String, n: String) = {
    val conf = new SparkConf().setMaster("local[*]").setAppName("task1")
    val sc = new SparkContext(conf)

    val rdd_review = sc.textFile(review_file).map(item => parse(item))
      .map(item => (compact(item \ "business_id"), compact(item \ "stars")))
      .filter(item => item._2 != null).map(item => (item._1, item._2.toDouble))
      .groupByKey().mapValues(item => item.toList)
    val rdd_business = sc.textFile(business_file).map(item => parse(item))
      .map(item => (compact(item \ "business_id"), compact(item \ "categories")))
      .filter(item => item._2 != null).filter(item => item._2 != "")
    //      .mapValues(item => item.split(",").toList)
    val rdd = rdd_business.join(rdd_review).map(item => (item._2._1, item._2._2))
      .flatMap(item => item._1.substring(1, item._1.length - 1).split(",").map(x => (x.trim(), item._2)))
      .reduceByKey((a, b) => a ::: b).mapValues(item => item.sum / item.size)
      .sortBy(item => item._1).sortBy(item => item._2, ascending = false)
//    print(rdd.take(n.toInt).toList)
    write_result(rdd.take(n.toInt).toList, output_file)
  }

  def t2_no_spark(review_file: String, business_file: String, output_file: String, n: String) = {
    val review = Source.fromFile(review_file)
    val business = Source.fromFile(business_file)

    val business_l = business.getLines().map(item => parse(item))
      .map(item => (compact(item \ "business_id"), compact(item \ "categories")))
      .filter(item => item._2 != null).filter(item => item._2.trim() != "")
    val review_l = review.getLines().map(item => parse(item))
      .map(item => (compact(item \ "business_id"), compact(item \ "stars").toDouble))
      .toSeq.groupBy(_._1).map(item => {
      var temp: List[Double] = List()
      item._2.foreach(i => {
        temp = temp :+ i._2
      })
      (item._1, temp)
    })

    val a = business_l.collect { case (k, v) if review_l.contains(k) => (v, review_l(k)) }
      .flatMap(item => {
        var l: List[(String, List[Double])] = List()
        item._1.substring(1, item._1.length - 1).split(",").foreach(i => {
          l = l :+ (i.trim(), item._2)
        })
        l
      }).map(item => (item._1, item._2.sum, item._2.size))
      .toSeq.groupBy(_._1).map(item => {
      var sum = 0.0
      var count = 0.0
      item._2.foreach(i => {
        sum += i._2
        count += i._3
      })
      (item._1, sum / count)
    }).toSeq.sortBy(_._1).reverse.sortBy(_._2).reverse
//    println(a.take(n.toInt).toList)
    write_result(a.take(n.toInt).toList, output_file)
  }

  def write_result(data:List[(String, Double)], output_file:String): Unit ={
    val pw = new PrintWriter(new File(output_file))
    pw.write("{\"result\": [")
    data.foreach(item => {
      pw.write("[\""+item._1+"\", "+item._2+"]")
      if (item._1 != data.last._1){
        pw.write(", ")
      }
    })
    pw.write("]}")
    pw.close()
  }

  def main(args: Array[String]): Unit = {


    // review_file, business_file, output_file, if_spark, n
    val review_file = args(0)//"E:\\Data\\inf553\\hw1\\review.json"
    val business_file = args(1)//"E:\\Data\\inf553\\hw1\\business.json"
    val output_file = args(2)//"task2_spark_ans"
    val if_spark = args(3)//"spark"
    val n = args(4)//"10"

    if (if_spark.trim() == "spark") {
      t2_spark(review_file, business_file, output_file, n)
    } else if (if_spark.trim() == "no_spark") {
      t2_no_spark(review_file, business_file, output_file, n)
    }


  }
}
